slrealizer/realize_om10.py

```diff
@@ -0,0 +1,298 @@
+        lens_mag = lensInfo[band + '_SDSS_lens']
+        lens_flux = from_mag_to_flux(lens_mag, to_unit='nMgy')
+        q_mag_arr = lensInfo[band + '_SDSS_quasar'] + from_flux_to_mag(np.abs(np.array(lensInfo['MAG'][:numQuasars])))
+        q_flux_arr = from_mag_to_flux(q_mag_arr, to_unit='nMgy')
+        q_tot_flux = np.sum(q_flux_arr)
+        derivedProps['apFlux'] = lens_flux + q_tot_flux
+        
+        #################################
+        # Analytical moment calculation #
+        #################################
+        
+        # Flux ratios weight the moment contributions from the lens and each quasar image
+        lensFluxRatio = lens_flux/derivedProps['apFlux']
+        qFluxRatios = q_flux_arr/derivedProps['apFlux']
+        
+        # Slice x, y positions to number of quasars
+        lensInfo['XIMG'] = lensInfo['XIMG'][:numQuasars]
+        lensInfo['YIMG'] = lensInfo['YIMG'][:numQuasars]
+        
+        # Convert Gaussian FWHM, HLR to sigma of the covariance matrix
+        sigmaSqPSF = fwhm_to_sigma(PSF_FWHM)**2.0
+        sigmaSqLens = hlr_to_sigma(lensInfo['REFF_T'])**2.0
+        
+        # Compute first moment contributions from the lens and each quasar image
+        lensIxContrib = lensIyContrib = 0.0 # because lens is centered at (0, 0)
+        qIxContrib = qFluxRatios*lensInfo['XIMG']
+        qIyContrib = qFluxRatios*lensInfo['YIMG']
+        derivedProps['x'] = lensIxContrib + np.sum(qIxContrib)
+        derivedProps['y'] = lensIyContrib + np.sum(qIyContrib)
+        
+        # Compute second moment contributions from the lens and each quasar image
+        lensIxxContrib = lensFluxRatio*(sigmaSqPSF + sigmaSqLens + derivedProps['x']**2.0)
+        lensIyyContrib = lensFluxRatio*(sigmaSqPSF + sigmaSqLens + derivedProps['y']**2.0)
+        lensIxyContrib = lensFluxRatio*derivedProps['x']*derivedProps['y'] # because lens is centered at (0, 0)
+        qIxxContrib = qFluxRatios*((lensInfo['XIMG'] - derivedProps['x'])**2.0 + sigmaSqPSF)
+        qIyyContrib = qFluxRatios*((lensInfo['YIMG'] - derivedProps['y'])**2.0 + sigmaSqPSF)
+        qIxyContrib = qFluxRatios*(lensInfo['XIMG'] - derivedProps['x'])*(lensInfo['YIMG'] - derivedProps['y'])
+        
+        # Add to get total second moments
+        Ixx = lensIxxContrib + np.sum(qIxxContrib)
+        Iyy = lensIyyContrib + np.sum(qIyyContrib)
+        Ixy = lensIxyContrib + np.sum(qIxyContrib)
+        derivedProps['trace'] = Ixx + Iyy
+        if self.DEBUG: derivedProps['det'] = Ixx*Iyy - Ixy**2.0
+        
+        # Compute ellipticities from second moments
+        denom = Ixx + Iyy #+ 2.0*(Ixx*Iyy - Ixy**2.0)**0.5 # uncommenting gives g1, g2 
+        derivedProps['e1'] = (Ixx - Iyy)/denom
+        derivedProps['e2'] = 2.0*Ixy/denom
+        
+        if self.DEBUG:
+            derivedProps['Ixx'] = Ixx
+            derivedProps['Iyy'] = Iyy
+            derivedProps['Ixy'] = Ixy
+        
+        return derivedProps
```

Comment:
Excellent! Did you write a unit test to check the moment calculations in some simple case? And maybe one to show that the analytical and galsim HSM moment calculations agree?

Reply:
Yes! Unit test at `tests/test_analytical_moments.py`

Reply:
As I mentioned earlier in one of our meetings, HSM was not a good method for multi-object systems because it uses an elliptical Gaussian. So I wrote methods in `slrealizer/utils/utils.py` that could take in an image and calculate the first and second moments (which I call the "raw numerical" method as opposed to the "hsm" method throughout the code). The analytical and the raw numerical methods agree.